tests/runner/test_run_node.py

```diff
@@ -0,0 +1,89 @@
+
+
+def generate_tuple():
+    for i in range(10):
+        yield i, i * i
+
+
+def generate_list():
+    for i in range(10):
+        yield [i, i * i]
+
+
+def generate_dict():
+    for i in range(10):
+        yield dict(idx=i, square=i * i)
+
+
+class TestRunGeneratorNode:
+    def test_generator_fail_async(self, mocker, catalog):
+        fake_dataset = mocker.Mock()
+        catalog.add("result", fake_dataset)
+        n = node(generate_one, inputs=None, outputs="result")
+
+        with pytest.raises(Exception, match="nodes wrapping generator functions"):
+            run_node(n, catalog, _NullPluginManager(), is_async=True)
+
+    def test_generator_node_one(self, mocker, catalog):
+        fake_dataset = mocker.Mock()
+        catalog.add("result", fake_dataset)
+        n = node(generate_one, inputs=None, outputs="result")
+        run_node(n, catalog, _NullPluginManager())
+
+        expected = [((i,),) for i in range(10)]
+        assert 10 == fake_dataset.save.call_count
+        assert fake_dataset.save.call_args_list == expected
+
+    def test_generator_node_tuple(self, mocker, catalog):
+        left = mocker.Mock()
+        right = mocker.Mock()
+        catalog.add("left", left)
+        catalog.add("right", right)
+        n = node(generate_tuple, inputs=None, outputs=["left", "right"])
+        run_node(n, catalog, _NullPluginManager())
+
+        expected_left = [((i,),) for i in range(10)]
+        expected_right = [((i * i,),) for i in range(10)]
+        assert 10 == left.save.call_count
+        assert left.save.call_args_list == expected_left
+        assert 10 == right.save.call_count
+        assert right.save.call_args_list == expected_right
+
+    def test_generator_node_list(self, mocker, catalog):
+        left = mocker.Mock()
+        right = mocker.Mock()
+        catalog.add("left", left)
+        catalog.add("right", right)
+        n = node(generate_list, inputs=None, outputs=["left", "right"])
+        run_node(n, catalog, _NullPluginManager())
+
+        expected_left = [((i,),) for i in range(10)]
+        expected_right = [((i * i,),) for i in range(10)]
+        assert 10 == left.save.call_count
+        assert left.save.call_args_list == expected_left
+        assert 10 == right.save.call_count
+        assert right.save.call_args_list == expected_right
+
+    def test_generator_node_dict(self, mocker, catalog):
+        left = mocker.Mock()
+        right = mocker.Mock()
+        catalog.add("left", left)
+        catalog.add("right", right)
+        n = node(generate_dict, inputs=None, outputs=dict(idx="left", square="right"))
+        run_node(n, catalog, _NullPluginManager())
+
+        expected_left = [((i,),) for i in range(10)]
+        expected_right = [((i * i,),) for i in range(10)]
+        assert 10 == left.save.call_count
+        assert left.save.call_args_list == expected_left
+        assert 10 == right.save.call_count
+        assert right.save.call_args_list == expected_right
```

Comment:
There is a good deal of repeated code in these three tests, what do you think about parametrising them?

```suggestion
    @mark.parametrize(
        "generator,outputs",
        [
            (generate_tuple, ["left", "right"]),
            (generate_list, ["left", "right"]),
            (generate_dict, dict(idx="left", square="right"))
         ]
     )
    def test_generator_nodes(self, mocker, catalog):
        left = mocker.Mock()
        right = mocker.Mock()
        catalog.add("left", left)
        catalog.add("right", right)
        n = node(generator, inputs=None, outputs=outputs)
        run_node(n, catalog, _NullPluginManager())

        expected_left = [((i,),) for i in range(10)]
        expected_right = [((i * i,),) for i in range(10)]
        assert 10 == left.save.call_count
        assert left.save.call_args_list == expected_left
        assert 10 == right.save.call_count
        assert right.save.call_args_list == expected_right
```

Reply:
I'd rather have them separately, keeping the tests DAMP is more important to me than having them DRY. Although there's different takes on that:
https://stackoverflow.com/a/11837973